src/core/qgsvectorlayer.cpp

```diff
@@ -2243,10 +2248,34 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
 
 bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue )
 {
-  if ( !mEditBuffer || !mDataProvider )
-    return false;
+  if ( fields().fieldOrigin( field ) == QgsFields::OriginJoin )
+  {
+    int srcFieldIndex;
+    const QgsVectorLayerJoinInfo *info = mJoinBuffer->joinForFieldIndex( field, fields(), srcFieldIndex );
+    if ( info && info->joinLayer() && info->isEditable() )
+    {
+      QgsFeature feature = getFeature( fid );
+
+      if ( !feature.isValid() )
+        return false;
+
+      QgsFeature joinFeature = mJoinBuffer->joinedFeatureOf( info, feature );
 
-  return mEditBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      if ( joinFeature.isValid() )
+        return info->joinLayer()->changeAttributeValue( joinFeature.id(), srcFieldIndex, newValue, oldValue );
```

Comment:
> I'm also curious how this approach would work with the use case of joined columns for label data defined x/y/ Wouldn't this use case require automatic switching on of edit mode for the joined table containing x/y columns?

Actually, as mentionned in the QEP : 

> Auxiliary data may be updated by the user without updating source data, or for read only source data.

So even if it's true we'll have to automatically switch the edit mode of the joined table, it should not be linked to the edit mode of the target layer (which may be read only).